### PR TITLE
4 register team

### PR DESCRIPTION
**Pull request by witterless**

### CODE REVIEW

**Review by GavinF17 — COMMENTED** [review]
Need to merge master :)

File: src/main/java/com/gavinfenton/quizolation/controller/TeamController.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    public TeamController(TeamService teamService) {
+        this.teamService = teamService;
+    }
+
+    @PostMapping(Endpoints.TEAMS)
+    public ResponseEntity<Team> createTeam(@RequestBody Team team, HttpServletRequest request) {
+        Team createdTeam = teamService.createTeam(team);
+
+        return ResponseEntity.created(URI.create(request.getRequestURI() + createdTeam.getId())).body(createdTeam);
```
Comment (GavinF17) [inline src/main/java/com/gavinfenton/quizolation/controller/TeamController.java]: Slash isn't appearing before the ID in location, seen it was for you so maybe your request URL already had it?

Screenshot from 2020-04-12 22-19-57

File: src/main/java/com/gavinfenton/quizolation/service/TeamService.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+public class TeamService {
+
+    private final TeamRepository teamRepository;
+
+    public TeamService(TeamRepository teamRepository) {
+        this.teamRepository = teamRepository;
+    }
+
+    public Team createTeam(Team team) {
+        return teamRepository.save(team);
```
Comment (GavinF17) [inline src/main/java/com/gavinfenton/quizolation/service/TeamService.java:21]: ID should be set to null, if it isn't then this will update rather than create a new one

File: src/main/java/com/gavinfenton/quizolation/entity/Quiz.java [hunk 3]
```diff
@@ -1,5 +1,6 @@
 package com.gavinfenton.quizolation.entity;
 
+import com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (GavinF17) [inline src/main/java/com/gavinfenton/quizolation/entity/Quiz.java]: Should be removed

File: src/main/java/com/gavinfenton/quizolation/service/QuizService.java [hunk 4]
```diff
@@ -26,13 +26,16 @@ public Quiz getQuiz(Long id) {
         return quizRepository.findById(id).orElseThrow(() -> new ObjectNotFoundException(id, "Quiz"));
     }
 
+    public List<Quiz> getAllQuizzes() {
+        return quizRepository.findAll();
+    }
```
Comment (GavinF17) [inline src/main/java/com/gavinfenton/quizolation/service/QuizService.java]: Merge issue, was renamed so shouldn't be there any more

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.verify;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: What's this for?

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    @MockBean
+    private TeamService teamService;
+
+    @BeforeEach
+    public void setUp() {
+        initMocks(this);
+    }
+
+    @Test
+    public void testGivenTeamsWhenTeamsThenReturnJsonArray() throws Exception {
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: Should have a clearer name

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
+package com.gavinfenton.quizolation.controller;
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java:1]: Can you add `//Given ... // When ... // Then` to all of these please? Also in the service tests

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(teamSaving));
+
+        // When
+        ResultActions response = mvc.perform(request);
+        Team teamActual = objectMapper.readValue(response.andReturn().getResponse().getContentAsString(), Team.class);
+
+        // Then
+        verify(teamService).createTeam(teamSaving);
+        response.andExpect(status().isCreated());
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java:90]: Add assertion for Location header

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        assertEquals(teamExpected, teamActual);
+    }
+
+    @Test
+    public void testGetTeasCallsAndReturnsTeamsFromService() throws Exception {
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: Teas :tea: 

19864642-0-image-a-116_1571358704234

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        Team teamExpected = new Team();
+
+        given(teamRepository.save(teamSaving)).willReturn(teamExpected);
+
+        Team teamActual = teamService.createTeam(teamSaving);
+
+        verify(teamRepository).save(teamSaving);
+        assertEquals(teamExpected, teamActual);
+    }
+
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java:48]: Add a test for create setting null when that is added in the service

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        Team teamActual = teamService.updateTeam(idExpected, teamSaving);
+        assertEquals(teamActual, teamExpected);
+    }
+
+    @Test
+    public void testGetTeam() {
+        Team teamSaving = new Team();
+        teamSaving.setId(1L);
+        teamSaving.setName("Team name");
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java:72]: Not saving anything so this entity isn't needed

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
+package com.gavinfenton.quizolation.service;
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java:1]: I would give more descriptive names of what is being tested as testCreateTeam() etc isn't descriptive, if functionality was added to it then it could be assumed that this is covered when it isn't

I think names like testCreateTeamSavesToAndReturnsFromRepository() lets you know that this is the only thing being tested (though it's a bit long)

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        teamSaving.setId(1L);
+        teamSaving.setName("Team name");
+        Long idExpected = teamSaving.getId();
+        Team teamExpected = new Team();
+
+        given(teamRepository.save(teamSaving)).willReturn(teamExpected);
+
+        Team teamActual = teamService.updateTeam(idExpected, teamSaving);
+        assertEquals(teamActual, teamExpected);
+    }
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java]: This is missing a test to make sure update sets the ID to the one passed in, I would use a separate test for this

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        verify(teamRepository).save(teamSaving);
+        assertEquals(teamExpected, teamActual);
+    }
+
+    @Test
+    public void testUpdateTeam() {
+        Team teamSaving = new Team();
+        teamSaving.setId(1L);
+        teamSaving.setName("Team name");
+        Long idExpected = teamSaving.getId();
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java]: Mixing saving and expected entities could pollute the tests, better to keep them separate so that you can be sure the one returned is from the repo and not the one passed in (i.e. could pass if update() does nothing but return)

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        Team teamExpected = new Team();
+
+        given(teamRepository.save(teamSaving)).willReturn(teamExpected);
+
+        Team teamActual = teamService.updateTeam(idExpected, teamSaving);
+        assertEquals(teamActual, teamExpected);
```
Comment (GavinF17) [inline src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java]: Add a verify() to make sure the repo was called

**Review by GavinF17 — CHANGES_REQUESTED** [review]

File: src/main/java/com/gavinfenton/quizolation/controller/TeamController.java (hunk 1 above)
Comment (witterless) [reply src/main/java/com/gavinfenton/quizolation/controller/TeamController.java]: You'd be right. Resolved.

File: src/main/java/com/gavinfenton/quizolation/service/TeamService.java (hunk 2 above)
Comment (witterless) [reply src/main/java/com/gavinfenton/quizolation/service/TeamService.java:21]: Resolved

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java (hunk 5 above)
Comment (witterless) [reply src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: Removed. Was for testing something when my tests failed.

Comment (GavinF17) [reply src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: Import just needs to be removed :)

File: src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java (hunk 9 above)
Comment (GavinF17) [reply src/test/java/com/gavinfenton/quizolation/controller/TeamControllerTest.java]: :tea: Still teas

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java (hunk 14 above)
Comment (GavinF17) [reply src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java]: Still mixing values of saving and expected, I would just remove the setting of ID completely as we don't need it for the test. Just pass in a magic number to update call in then().

Creating two Teams and changing the name of one will make sure that saving and expected don't equals() each other, which is what makes the assertEquals(teamActual, teamExpected); valid as they could only equal if the repo returned teamExpected and not teamSaving (as it has a different name looked at in equals())

File: src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java (hunk 11 above)
Comment (GavinF17) [reply src/test/java/com/gavinfenton/quizolation/service/TeamServiceTest.java:72]: You copy & pasted the teams test for this, so getTeam() isn't being tested any more

**Review by GavinF17 — APPROVED** [review]
